### PR TITLE
Fixes ad conversions fail after a creative has expired

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/10240

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

- Confim ad conversions work as expected after the creative has been removed from the catalog, i.e. expired
- Confirm ad conversions are removed from the `Default/ads_service/bundfle_state` sqlite database when the expiry date has elapsed (see `expiry_timestamp` column)

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/catalog_state.cc [hunk 1]
```diff
@@ -132,6 +132,17 @@ Result CatalogState::FromJson(
         ad_conversion.observation_window =
             conversion["observationWindow"].GetUint();
 
+        base::Time end_at_timestamp;
+        if (!base::Time::FromUTCString(campaign_info.end_at.c_str(),
+            &end_at_timestamp)) {
+          BLOG(1, "FOOBAR.2: " << campaign_info.end_at);
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/catalog_state.cc:138]: for debugging?

File: components/brave_ads/browser/bundle_state_database.cc [hunk 2]
```diff
@@ -342,24 +342,29 @@ bool BundleStateDatabase::CreateAdConversionsTable() {
...
-          "creative_set_id LONGVARCHAR NOT NULL, "
+          "(creative_set_id LONGVARCHAR NOT NULL, "
           "type LONGVARCHAR NOT NULL, "
           "url_pattern LONGVARCHAR NOT NULL, "
-          "observation_window INTEGER NOT NULL)",
+          "observation_window INTEGER NOT NULL, "
+          "expiry_timestamp TIMESTAMP, "
+          "UNIQUE(creative_set_id, type, url_pattern, "
+              "observation_window), "
+          "PRIMARY KEY(creative_set_id, type, url_pattern, "
```
Comment (moritzhaller) [inline components/brave_ads/browser/bundle_state_database.cc:352]: should `expiry_timestamp` be part of the PK to allow for a new conversion with an updated expiry date to be saved in the case that a campaign date updated in flight?

File: vendor/bat-native-ads/src/bat/ads/internal/catalog_state.cc (hunk 1 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/catalog_state.cc:138]: Fixed :-)

File: components/brave_ads/browser/bundle_state_database.cc (hunk 2 above)
Comment (tmancey) [reply components/brave_ads/browser/bundle_state_database.cc:352]: do we want duplicate entries with different timestamps and observation windows?

**Comment by tmancey** [comment]
CI failed on test-install

**Comment by kjozwiak** [comment]
Verification **PASSED** on `macOS 10.15.5 x64` using the following build:

```
Brave| 1.12.61 Chromium: 83.0.4103.116 (Official Build) nightly (64-bit)
--- | ---
Revision | 8f0c18b4dca9b6699eb629be0f51810c24fb6428-refs/branch-heads/4103@{#716}
OS | macOS Version 10.15.5 (Build 19F101)
```

* launched brave using
* viewed `https://www.travala.com` and closed Brave
* removed the `creativeInstanceId` from `catalog.json` that belonged to the ad that was viewed
* relaunched brave and visited `https://www.travala.com/booking?check_in=*`
* ensured that the conversion timer for the creative set started even though the `creativeInstanceId` was removed from `catalog.json`

```
[76939:775:0630/110006.110784:VERBOSE1:ad_conversions.cc(63)] Checking URL for ad conversion
[76939:775:0630/110006.148584:VERBOSE1:ads_impl.cc(751)] Classified page as science-science. Winning page classification over time is technology & computing-software
[76939:775:0630/110006.149657:VERBOSE1:ad_conversions.cc(129)] Ad conversion for creative set id 37f7ba07-355d-4d0f-8cdb-b9241550c0c8 and postview
[76939:775:0630/110006.149810:VERBOSE3:ad_conversions.cc(301)] Saving ad conversions state
[76939:775:0630/110006.149991:VERBOSE1:ad_conversions.cc(293)] Started ad conversion timer for creative instance id cd8a81fc-5efb-41f1-99b6-945082c3e83a and creative set id 37f7ba07-355d-4d0f-8cdb-b9241550c0c8 which will trigger in 13 hours, 1 minute, 29 seconds at 12:01 AM
```

* closed brave and moved desktop ~1 day ahead so the ad coversion triggers

```
[77068:775:0701/110645.249810:VERBOSE1:ad_conversions.cc(293)] Started ad conversion timer for creative instance id cd8a81fc-5efb-41f1-99b6-945082c3e83a and creative set id 37f7ba07-355d-4d0f-8cdb-b9241550c0c8 which will trigger in 0 hours, 2 minutes, 14 seconds at 11:08 AM
```

* waited ~2-3 minutes

```
[77068:775:0701/110859.253623:VERBOSE1:ad_conversions.cc(253)] Successfully converted ad with creative instance id cd8a81fc-5efb-41f1-99b6-945082c3e83a and creative set id 37f7ba07-355d-4d0f-8cdb-b9241550c0c8 in -11 hours, -7 minutes, -23 seconds at 12:01 AM
[77068:775:0701/110859.253990:VERBOSE3:ad_conversions.cc(301)] Saving ad conversions state
[77068:775:0701/110859.254087:VERBOSE1:ad_conversions.cc(78)] Ad conversion queue is empty
```

```
[77068:775:0701/110900.290894:VERBOSE3:confirmations_impl.cc(750)] Saving confirmations state
[77068:775:0701/110900.291337:VERBOSE1:confirmations_impl.cc(1229)] Successfully redeemed unblinded token with confirmation id dc8b1e60-8c65-4d91-9f5a-2f262c00e1c3, creative instance id cd8a81fc-5efb-41f1-99b6-945082c3e83a and conversion
```

**Pull request merged**